include/sys/dbuf.h

```diff
@@ -442,10 +445,34 @@ dbuf_find_dirty_eq(dmu_buf_impl_t *db, uint64_t txg)
 	(dbuf_is_metadata(_db) &&					\
 	((_db)->db_objset->os_primary_cache == ZFS_CACHE_METADATA)))
 
-#define	DBUF_IS_L2CACHEABLE(_db)					\
-	((_db)->db_objset->os_secondary_cache == ZFS_CACHE_ALL ||	\
-	(dbuf_is_metadata(_db) &&					\
-	((_db)->db_objset->os_secondary_cache == ZFS_CACHE_METADATA)))
+/*
+ * We want to exclude buffers that are on a special allocation class from
+ * L2ARC.
+ */
+static inline boolean_t dbuf_is_l2cacheable(dmu_buf_impl_t *db)
+{
+	blkptr_t *bp = db->db_blkptr;
+	spa_t *spa = db->db_objset->os_spa;
+	uint64_t vdev = DVA_GET_VDEV(db->db_blkptr->blk_dva);
```

Comment:
Accessing `db->db_blkptr` before the `bp != NULL` check.  It looks like you can move this in to the `if (bp != NULL) { }` block